ROADMAP.md

```diff
@@ -0,0 +1,89 @@
+# Chisel Roadmap
+
+This roadmap captures the current plans of the Chisel maintainers and can be
+changed through the usual PR process.
+Not every small feature add or planned bugfix is logged here, but the high-level
+changes and planned direction that are relevant for those using and contributing
+to the project.
+Two especially notable directions on the roadmap below are:
+
+* We are migrating Chisel to rely on, exclusively, the
+  [MLIR-based FIRRTL Compiler](https://github.com/llvm/circt)
+* Accelerating improvements and changes to Chisel and the FIRRTL specification
+
+
+## 3.5.x (and Earlier)
+
+- Continue to maintain binary compatibility on these releases
+- Continue to backport bug fixes, new features if possible, documentation
+  updates, etc.
+- Deprecations are backported to ease in users migrating to next major release
+  (3.6).
+
+## 3.6 (Target 2022 November)
+
+This release is intended to provide a smooth path for the Chisel codebase to no
+longer depend on the 
+[Scala FIRRTL Compiler (SFC)](https://github.com/chipsalliance/firrtl)
+and use the [MLIR Firrtl Compiler (MFC)](https://github.com/llvm/circt) as the
+main backend. While the SFC may continue to work to consume the `.fir` produced
+by Chisel, it will not be a requirement for adding new features to the FIRRTL
+spec and the Chisel APIs. Thus, the 3.6 release will:
+
+- Deprecate `Chisel._`
+- Turn the deprecations from previous releases into Errors
+- Support integration with [MFC](https://github.com/llvm/circt) by bringing
+  `chisel-circt` in-repo
+- Stop using [SFC](https://github.com/chipsalliance/firrtl) except to test
+  deprecated features
+- Deprecate all Chisel features that are not implemented in MFC. A nonexhaustive
+  list includes:
+  - Interval Types
+  - Fixed Point Types
+  - Injecting Aspects
+  - EnumAnnotations
+  - RunFirrtlTransformAnnotation
+  - LoadMemoryFromFile
+  - Trace API
+  - ChiselAnnotation/custom annotations
```

Comment:
if there is no support for the decoder API, does that mean that rocket-chip won't be able to update?

Reply:
The decoder is resolved in the front end, they are pure chisel api, which has been masked in the mfc. 
the reason I added them is: if users are sticking on qmc decoder, they can use this api to cache the decoder results to speed up elaboration among different runs.

Reply:
The plan, and current implementation of MFC, is to reject any annotations that it does not explicitly understand. (This may be disabled by a user, but almost always without the correct effect on their design.) A `chisel-circt` based flow can output a Chisel `.anno.json` which includes things like decoder results. This doesn't have to be ingested/understood by a FIRRTL compiler.

Reply:
i think the question about EnumAnnotations  would match this as well

Reply:
Clarifying: `ChiselAnnotation` will be made a `private[chsiel3]` API and is thereby deprecated for Chisel users. Any custom annotation, i.e., anything not in-tree with Chisel3, will not be supported. That said, we're very, very open to improvements to Chisel, the FIRRTL spec, and the MFC to meet user needs, e.g., enumerated types in FIRRTL to replace enum annotations, as well as continuing to seamlessly support Chisel APIs that are currently emitted using annotations, e.g., trace, inline/flatten, and dontTouch (_these examples are non-exhaustive_). We'll address FireSim usages in the near future.

Reply:
Simply put, we are totally missing a good pass injection flow for the MFC. That needs to be added. You should be able to, LLVM-style, inject a custom pass built as a shared library into MFC and have it work.

The main question is then what to do with custom annotations. Thinking about this a bit more... what we could support is a true "custom annotation" which requires a user to opt-in and only supports a `SingleTargetAnnotation` format (all `MultiTargetAnnotations` can be converted to a formulation that uses multiple `SingleTargetAnnotations`). There are already kind-of hooks for this in CIRCT that exist for testing only with fake annotations:

1) `circt.test` is a `SingleTargetAnnotation` that can be non-local
2) `circt.testLocalOnly` is a `SingleTargetAnnotation` that _must_ be local (effectively point at a `Named`)
3) `circt.testNT` is a `NoTargetAnnotation`

These _could_ be enshrined as a special annotation format called `CustomAnnotation.TargetLocal` and `CustomAnnotation.TargetNonLocal` (where a `NoTargetAnnotation` has to be an annotation that targets the circuit). This then enables arbitrary custom annotations to be shoved into CIRCT. We currently have an optional warning that annotations must be removed by the time `LowerToHW` (conversion of FIRRTL Dialect to HW Dialects happens). If we enable custom annotations, then this should be an error by default and demotable to a warning.

An example of the above (using existing infra) is:

```
circuit Foo: %[[
  {
    "class": "circt.test",
    "target": "~Foo|Foo>w",
    "payload": {
      "class": "gcd.tools.ResilientRegisterAnnotation"
    }
  }
]]
  module Foo:
    input a: UInt<1>
    output b: UInt<1>

    wire w: UInt<1>
    w <= a
    b <= w
```

Producing:

```mlir
module {
  firrtl.circuit "Foo"  {
    firrtl.module @Foo(in %a: !firrtl.uint<1>, out %b: !firrtl.uint<1>) {
      %w = firrtl.wire interesting_name  {annotations = [{class = "circt.test", payload = {class = "gcd.tools.ResilientRegisterAnnotation"}}]} : !firrtl.uint<1>
      firrtl.strictconnect %w, %a : !firrtl.uint<1>
      firrtl.strictconnect %b, %w : !firrtl.uint<1>
    }
  }
}
```

We then just need at least two hooks for custom passes: 

1) You need a pass right after `LowerAnnotations` that can resolve the annotations to something else.
2) You need a pass somewhere else in the FIRRTL pipeline to remove the annotations.

My main concern about this approach is that mid-term, almost all annotations (which are not representable using the `TraceAPI`) are really custom operations and _not_ existing operations with discarable metadata. There is nothing in CIRCT that can know what is the behavior of a custom annotation is nor should it know. Put differently, a TMR Register is better modeled as `firrtl.reg.tmr` as opposed to `firrtl.reg annotation {"tmr"}`. It would be better, mid-term, if we worked towards enabling Chisel to construct MLIR directly which could enable novel, library-linked FIRRTL or other dialect operations to be constructed. This seems possible, but I don't have line-of-sight on how to do it.

For your second point of scan chain connection. This is really something that should be handled on the Chisel side. @jackkoenig has some experiments with providing a resource, having resources add information about what they need to do, and then having the resource make modifications to the module. More concretely, a scan chain could work like:

```scala
class ScanChainResource {
  def addToScanChain[A <: Data](a: Data): ???

  def buildScanChain(): Unit
}

class Bar(scanChain: ScanChainResource) extends Module {
  val r = Reg(Bool())
  scanChain.addToScanChain(r)
}

class Foo extends Module {
  val scanChain = new ScanChainResource

  val bar1, bar2 = Module(new Bar(scanChain))

  scanChain.buildScanChain()
}
```

Or: what Albert was doing with the scan chain stuff (and [what I've wrote passes for in the past](https://github.com/IBM/chiffre/blob/master/src/main/scala/chiffre/passes/ScanChainTransform.scala) 😬 ) is really a missing first-class Chisel ability. We didn't have this, so we always fell back to using the FIRRTL compiler in somewhat inappropriate ways.

WDYT?

Reply:
This sounds like a great path forward. FWIW the TMR and scan-chain passes were just examples, but I'm interested in supporting other transforms that we use internally, and I'm sure other companies have plenty of transforms that they'd like to trigger with custom annotations.

Regarding the annotations that we pass on from Chisel to the MFC, it seems like two pieces of information are generally important:

- What is the target of the annotation (some circuit element/module, either local or non-local)
- What is the "type" of the annotation

It looks like from your example, that "payload.class" field gives that second piece of information. I haven't seen that sort of pattern in Chisel/CIRCT yet, but it looks like a great way to get this information across.

And agreed, in the not-immediate term I really like the idea of extending the FIRRTL MLIR dialect in some library-linked manner, but I have no idea how that sort of thing would work out-of-tree from the CIRCT repo.